torch_geometric/nn/models/label_usage.py

```diff
@@ -0,0 +1,92 @@
+import torch
+import torch.nn.functional as F
+
+from torch_geometric.typing import Adj, OptTensor
+
+
+class LabelUsage(torch.nn.Module):
+    r"""The label usage operator for semi-supervised node classification,
+    as introduced in `"Bag of Tricks for Node Classification"
+    <https://arxiv.org/abs/2103.13355>`_ paper.
+
+    .. note::
+
+        When using the :class:`LabelUsage`, adjust the model's input dimension
+        accordingly to include both features and classes.
+
+    Args:
+        base_model: An instance of the model that will do the
+            inner forward pass.
+        num_classes (int): Number of classes in dataset
+        split_ratio (float): Proportion of true labels to use as features
+            during training (default: :obj:'0.5')
+        num_recycling_iterations (int): Number of iterations for the
+            label reuse procedure to cycle predicted soft labels
+            (default: :obj:'100')
+        return_tuple (bool): If true, returns (output, train_label_idx,
+            train_pred_idx) otherwise returns prediction output
+            (default :obj:'False')
+    """
+    def __init__(
+        self,
+        base_model: torch.nn.Module,
+        num_classes: int,
+        split_ratio: float = 0.5,
+        num_recycling_iterations: int = 100,
+        return_tuple: bool = False,
+    ):
+
+        super().__init__()
+        self.base_model = base_model
+        self.num_classes = num_classes
+        self.split_ratio = split_ratio
+        self.num_recycling_iterations = num_recycling_iterations
+        self.return_tuple = return_tuple
+
+    def forward(
+        self,
+        x: Tensor,
+        edge_index: Adj,
+        y: Tensor,
+        train_idx: Tensor,
+        batch: OptTensor = None,
+    ):
+        r"""Forward pass using label usage algorithm.
+
+        Args:
+            x (torch.Tensor): Node feature tensor
+            edge_index (torch.Tensor or SparseTensor): The edge indices
+            y (torch.Tensor): Node label tensor
+            train_idx (torch.Tensor): Global indices of all
+                nodes in training set
+            batch (torch.Tensor, optional): Optional mini-batch specification
+                indicating global node indices in batch (default: :obj:'None')
+        """
+        # re-assign train_idx to be nodes in mini-batch
+        if batch is not None:
+            train_idx = batch
```

Comment:
batch is never used again, so I'm not sure it makes sense to have both batch and train_idx parameters.
@liuvince is current impl what you had in mind for minibatches? In the case of minibatching I think x and y would just not be the whole dataset.

Reply:
I can remove batch and indicate that train_idx can be used for batch indices since batches passed contain the global indices. That seems more intuitive now that I read your comment and just from looking through it.

Reply:
Can we name it `mask` to ensure aligned naming with https://pytorch-geometric.readthedocs.io/en/2.5.2/generated/torch_geometric.nn.models.LabelPropagation.html and https://pytorch-geometric.readthedocs.io/en/2.5.2/generated/torch_geometric.nn.models.CorrectAndSmooth.html for example ?

Reply:
I propose the following logic:
in training mode (when `self.training==True`):
- when `mask` is set to None, we should perform the split in the forward pass.
- when `mask` is set to not-null value, we assume that it corresponds to nodes which labels are used as node features.
in test mode:
-  when `mask` is set to None, we use the whole input as unlabeled data.
- when `mask` is set to not-null value, we assume that it corresponds to nodes which labels are used as node features which is the whole or sampled training dataset.
does it make sense ?

Reply:
Yeah I can rename it to `mask` for keeping with convention

Reply:
Also not sure it makes sense to allow None for mask.  In both training and evaluation we need to know which set of nodes its ok to use true labels for.  

Can we update the `mask` description to be more similar to the examples Vincent linked?
Something like "A mask or index tensor denoting which nodes true labels can be used. "

As is it's a little confusing because it's called "mask" but we describe it as an index tensor.  A mask is generally a bool tensor of shape (N,) while an index tensor is a list of the indices where you would put "true" in the mask, so it would have shape (k,) where k<=N.  We can skip this detail on the expected size for this parameter since it is a common concept.  I think the code should work for both cases without any changes.